otter/controller.py

```diff
@@ -467,3 +473,44 @@ def convergence_remove_server_from_group(
         yield do_return(assoc_obj(state, desired=state.desired - 1))
     else:
         yield do_return(state)
+
+
+def remove_server_from_group(log, trans_id, server_id, replace, purge,
+                             group, state, config_value=config_value):
+    """
+    Remove a specific server from the group, optionally replacing it
+    with a new one, and optionally deleting the old one from Nova.
+
+    If the old server is not deleted from Nova, otter-specific metadata
+    is removed: otherwise, a different part of otter may later mistake
+    the server as one that *should* still be in the group.
+
+    :param log: A bound logger
+    :param bytes trans_id: The transaction id for this operation.
+    :param bytes server_id: The id of the server to be removed.
+    :param bool replace: Should the server be replaced?
+    :param bool purge: Should the server be deleted from Nova?
+    :param group: The scaling group to remove a server from.
+    :type group: :class:`~otter.models.interface.IScalingGroup`
+    :param state: The current state of the group.
+    :type state: :class:`~otter.models.interface.GroupState`
+
+    :return: The updated state.
+    :rtype: deferred :class:`~otter.models.interface.GroupState`
+    """
+    # worker case
+    if not tenant_is_enabled(group.tenant_id, config_value):
+        return worker_remove_server_from_group(
+            log, trans_id, server_id, replace, purge, group, state)
+
+    def kick_off_convergence(new_state):
+        cs.start_convergence(log, group.tenant_id, group.uuid)
+        return new_state
+
+    # convergence case - requires that the convergence dispatcher handles
+    # EvictServerFromScalingGroup
+    cs = get_convergence_starter()
```

Comment:
There's quite a lot going on in this function. It looks like the test you wrote is pretty good, but I wonder if splitting it up would make it slightly easier. I generally prefer perform() to be in tail-call position of a function, that way no one function is dealing with both Effects and Deferreds, which can be a bit confusing. (I broke this rule in `ConvergenceStarter`, but other things that call perform() in `service.py` do follow the rule. In `ConvergenceStarter`, the only thing done with the result is adding a logging callback).

However, I will leave it up to your discretion, because I'm not strongly bothered by it. This PR LGTM if you want to merge it now. :+1: